Add last changed by for ozw lock

zyberzero opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This will set the changed by attribute to the User Code attribute provided by the Zwave-lock, if it exists.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

zyberzero committed:
OZW: Add Last Changed By for Lock device

MartinHjelmare commented on homeassistant/components/ozw/lock.py:
```diff
@@ -74,6 +75,19 @@ def is_locked(self):
+        if user_code is None:
+            return
+
+        return f"{user_code.value}"
```
Isn't the user code a credential? We don't set credentials in the state machine.

zyberzero replied:
> Isn't the user code a credential? We don't set credentials in the state machine.

No, OZW calls it the user code but it is more like the slot ID for a user. What's stored in USER_CODE is the Slot ID of the user, not the credential itself.

balloob replied:
Why do you convert it to a string?

MartinHjelmare replied:
According to our dev docs this property should return a string. If the values is not a string from the beginning its good to copy it.

Whats the native value type?

https://developers.home-assistant.io/docs/core/entity/lock#properties

zyberzero replied:
https://github.com/cgarwood/python-openzwave-mqtt/blob/master/openzwavemqtt/models/value.py#L28
The return type of OZWNode.value is Any - how ever, what's sent on the MQTT topic for this value is a byte.
```
{
    "Label": "User Code",
    "Value": 0,
    "Units": "",
    "ValueSet": true,
    "ValuePolled": false,
    "ChangeVerified": false,
    "Min": 0,
    "Max": 255,
    "Type": "Byte",
    "Instance": 1,
    "CommandClass": "COMMAND_CLASS_NOTIFICATION",
    "Index": 260,
    "Node": 49,
    "Genre": "User",
    "Help": "User Code ID that was used",
    "ValueIDKey": 73183494772899857,
    "ReadOnly": true,
    "WriteOnly": false,
    "Event": "valueChanged",
    "TimeStamp": 1611850845
}
```

So I think this is the proper way, maybe not the most pythonic way though.

MartinHjelmare replied:
Ok, then let's keep the string formatting.

MartinHjelmare commented on homeassistant/components/ozw/lock.py:
```diff
@@ -74,6 +75,19 @@ def is_locked(self):
+        )
+
+        if user_code is None:
+            return
```
```suggestion
            return None
```

MartinHjelmare reviewed (changes requested):
Please remove the empty dump.csv file and address the comments. Then we can merge I think.

MartinHjelmare commented:
Probably need to rebase the branch on latest dev branch too.

zyberzero committed:
return None if no user code is found

zyberzero commented:
Rebased on latest dev, fixed your comments @MartinHjelmare . Do you think that I should change @balloob 's comment as well? Ie, not explicity return the value as a string?

MartinHjelmare commented:
Please add a test that covers the added code.

MartinHjelmare commented on homeassistant/components/ozw/lock.py:
```diff
@@ -74,6 +75,19 @@ def is_locked(self):
+        )
+
+        if user_code is None:
+            return None
```
We need to cover this case too.
